### DISCUSSION TITLE
PWA app containerizing using docker

### DISCUSSION THREAD

**Discussion by maniraj1234** [DuCanhGH/next-pwa#123] [Q&A]
Can someone share any ref / example for any dockerized PWA? 
My pwa is working fine in dev but not showing up in docker. [may be some issue with docker file or any setting]
it would be great if anyone can share some example.

**Comment by DuCanhGH** [comment]
I mean, it is the same as https://github.com/vercel/next.js/tree/canary/examples/with-docker with `next-pwa` installed as long as you actually serve the service worker file where `next-pwa` expects it to be ("/sw.js" by default). By default (starting from v10), it is output to "public/sw.js" (meaning that it is accessible at "/sw.js"), so if you do copy the "public" folder to your dockerized Next.js app, it should work just fine.

**Reply by maniraj1234** [reply]
Yes, I'm using same image. infact, I could see the assets & sw files in docker container: 
![image](https://github.com/DuCanhGH/next-pwa/assets/26603468/e8c6f9c4-1bd4-4d2a-acc6-6cd9196cc050)

**Reply by DuCanhGH** [reply]
@maniraj1234 what problems are you having then? Is "/sw.js" accessible? Can I see your next.config.js? Thanks!

**Reply by maniraj1234** [reply]
```/** @type {import('next').NextConfig} */
const nextConfig = {
	output: "standalone",
	experimental: {
		instrumentationHook: true,
	},
	logging: {
		fetches: {
			fullUrl: true,
		},
	}
};

const withPWA = require("@ducanh2912/next-pwa").default({
	dest: "public",
	cacheOnFrontEndNav: true,
	aggressiveFrontEndNavCaching: true,
	reloadOnOnline: true,
	swcMinify: true,
	disable: false,
	workboxOptions: {
		disableDevLogs: true,
	}
  });
 
module.exports = withPWA(nextConfig)
```

it doesn't show up the PWA icon in browser from docker app. from dev app, it is showing that.

**Reply by DuCanhGH** [reply]
@maniraj1234 can you check whether visiting "/sw.js" resolves to a JavaScript file? If it does, also check if the service worker is activated :) There might have been some issues with its activation.

**Reply by maniraj1234** [reply]
It was accessible. how do I check if service worker is activated? no errrors on console though.

**Reply by DuCanhGH** [reply]
@maniraj1234 go to Application -> Service Workers and check if there is any service worker registered.

**Reply by maniraj1234** [reply]
Yes, that is registered too.  
![image](https://github.com/DuCanhGH/next-pwa/assets/26603468/c4b432d5-5376-4807-87d6-b56a52b59078)

**Reply by DuCanhGH** [reply]
@maniraj1234 looks like the service worker is working just fine. Did you add "/manifest.json"? It is essential to having that install button on your address bar.

Edit: forgot that you said it was working in dev. But why would the prod build be different to the dev build? Can you provide a reproduction?

**Reply by maniraj1234** [reply]
Manifest.json is present too:
![image](https://github.com/DuCanhGH/next-pwa/assets/26603468/c40a5f5d-20e4-4f18-a1d3-26de92e7758b)

in dev build, I didn't use docker, simple `npm run dev` but for prod build, I'm using docker.

**Reply by DuCanhGH** [reply]
@maniraj1234 well, the environment doesn't matter anymore. All the files are served correctly, so that can't be the issue. If you want to double check, perhaps you can try using the good old `next build && next start`? Also, can you try running Lighthouse to investigate further?

**Reply by maniraj1234** [reply]
lighthouse is saying no service worker is registered :(

**Reply by DuCanhGH** [reply]
@maniraj1234 1) what

Have you got a reproduction? These clues won't get me anywhere I'm afraid.

**Reply by maniraj1234** [reply]
Thank you for your valuable time. I gave you access to the repo: changes are there in this branch [mmadishetty/deployments/docker] if you can take a look at high level.

**Reply by DuCanhGH** [reply]
@maniraj1234 this is your pages/sales.js:
```js
import "../app/globals.css";
import MainComponent from "@/components/MainComponent";
import Navbar from "@/components/Navbar";

function SalesPage() {
	return (
		<div>
			<Navbar />
			<MainComponent />;
		</div>
	);
}

export default SalesPage;
```
You have to add `<Head><link rel="manifest" href="https://example.com/manifest.json" /></Head>` to this page, pages/_app.tsx, or pages/_document.tsx (preferably this one) for the manifest to be detected and consumed by the browser. Adding `manifest: "/manifest.json"` to app/layout.tsx's metadata doesn't fix this for you because it only applies for App Router's routes. Do understand the difference between Pages Router and App Router (I'd say that this might come as a little bit confusing to people new to Next.js's system of two routers). Hope this helps :)

I'd recommend migrating to App Router though, since there's only one route in the pages directory as far as I can see.

**Reply by maniraj1234** [reply]
Thank you very much for your time. This resolved the issue. this was a bit confusing as I'm new to nextJS.

**Reply by DuCanhGH** [reply]
@maniraj1234 yes, but this code
![image](https://github.com/DuCanhGH/next-pwa/assets/75556609/a26e26f9-fac3-467d-9762-71454a4f1c2e)
is wrong. You have to remove `<html />`. This is because in Pages Router, Next.js adds it for you automatically.

**Reply by maniraj1234** [reply]
thank you!
updated it to:
```<div>
<Head><link rel="manifest" href="/manifest.json" /></Head>
<Navbar />
<MainComponent />;
</div>```
